test/ParameterStoreBuildWrapperPluginTest.groovy

```diff
@@ -25,6 +33,154 @@ class ParameterStoreBuildWrapperPluginTest {
+        }
+    }
+
+    public class Apply {
+        @After
+        public void reset() {
+            Jenkinsfile.instance = null
+            ParameterStoreBuildWrapperPlugin.reset()
+        }
+
+        private configureJenkins(Map config = [:]) {
+            Jenkinsfile.instance = mock(Jenkinsfile.class)
+            when(Jenkinsfile.instance.getStandardizedRepoSlug()).thenReturn(config.repoSlug)
+            when(Jenkinsfile.instance.getRepoName()).thenReturn(config.repoName ?: 'repo')
+            when(Jenkinsfile.instance.getOrganization()).thenReturn(config.organization ?: 'org')
+            when(Jenkinsfile.instance.getEnv()).thenReturn(config.env ?: [:])
+        }
+
+        @Test
+        void doesNotDecorateTheTerraformValidateStageIfGlobalParametersNotSet() {
+            def expectedClosure                     = { -> }
+            TerraformValidateStage stage            = mock(TerraformValidateStage.class)
+            ParameterStoreBuildWrapperPlugin plugin = spy(new ParameterStoreBuildWrapperPlugin())
+
+            plugin.apply(stage)
+
+            verify(stage, never()).decorate(expectedClosure)
+        }
+
+        @Test
+        void decorateTheTerraformValidateStageIfGlobalParametersSet() {
+            String path                             = '/somePath/'
+            def expectedClosure                     = { -> }
+            Map gp                                  = [path: path]
+            TerraformValidateStage stage            = mock(TerraformValidateStage.class)
+            ParameterStoreBuildWrapperPlugin plugin = spy(new ParameterStoreBuildWrapperPlugin())
+
+            doReturn(expectedClosure).when(plugin).addParameterStoreBuildWrapper(gp)
+
+            plugin.withGlobalParameter(path)
+            plugin.apply(stage)
+
+            verify(stage).decorate(TerraformValidateStage.ALL, expectedClosure)
+        }
+
+        @Test
+        void decorateTheTerraformEnvironmentStageIfGlobalParametersNotSet() {
+            String organization = "MyOrg"
+            String repoName     = "MyRepo"
+            String environment  = "MyEnv"
+            Map apo             = [path: "/${organization}/${repoName}/${environment}/", credentialsId: "${environment.toUpperCase()}_PARAMETER_STORE_ACCESS"]
+            def expectedClosure = { -> }
+            configureJenkins(repoName: repoName, organization: organization)
+
+            TerraformEnvironmentStage stage         = mock(TerraformEnvironmentStage.class)
+            ParameterStoreBuildWrapperPlugin plugin = spy(new ParameterStoreBuildWrapperPlugin())
+
+            doReturn(environment).when(stage).getEnvironment()
+            doReturn(apo).when(plugin).getEnvironmentParameterOptions(environment)
+            doReturn(expectedClosure).when(plugin).addParameterStoreBuildWrapper(apo)
+
+            plugin.apply(stage)
+
+            verify(stage, times(2)).decorate(anyString(), eq(expectedClosure))
+        }
```

Comment:
The intent behind this test is a little confusing, and I think  the  problem is being  made more difficult with the extra setup that's not relevant to testing the `apply` method?

Reply:
It's worth nothing, that even with these changes to simplify, your previous tests have uncovered another "smell" that existed before you even picked up this PR.  Does `addParameterStoreBuildWrapper` really need to be called twice?

```
    public void apply(TerraformEnvironmentStage stage) {
        def environment = stage.getEnvironment()	        
        def options << getParameterOptions(environment)

        options.each { option ->
            def buildWrapperClosure = addParameterStoreBuildWrapper(option)
            stage.decorate(PLAN, buildWrapperClosure)
            stage.decorate(APPLY, buildWrapperClosure)
        }
```

Now your test would have even less to validate.